src/irmin-pack/irmin_pack_layers.ml

```diff
@@ -743,65 +739,46 @@ struct
           we have to iter over the graph of commits, iter over the graph of
           nodes, and lastly iter over contents. *)
       let copy_newies_aux ~with_lock t =
-        Log.debug (fun l -> l "copy newies");
         let newies_commits =
           if with_lock then
-            X.Commit.CA.unsafe_consume_newies () |> List.rev |> Lwt.return
-          else X.Commit.CA.consume_newies t.X.Repo.commit >|= List.rev
+            X.Commit.CA.unsafe_consume_newies t.X.Repo.commit |> Lwt.return
+          else X.Commit.CA.consume_newies t.X.Repo.commit
         in
         let newies_nodes =
           if with_lock then
-            X.Node.CA.unsafe_consume_newies () |> List.rev |> Lwt.return
-          else X.Node.CA.consume_newies t.X.Repo.node >|= List.rev
+            X.Node.CA.unsafe_consume_newies t.X.Repo.node |> Lwt.return
+          else X.Node.CA.consume_newies t.X.Repo.node
         in
         let newies_contents =
           if with_lock then
-            X.Contents.CA.unsafe_consume_newies () |> List.rev |> Lwt.return
-          else X.Contents.CA.consume_newies t.X.Repo.contents >|= List.rev
+            X.Contents.CA.unsafe_consume_newies t.X.Repo.contents |> Lwt.return
+          else X.Contents.CA.consume_newies t.X.Repo.contents
         in
-        let copy_contents contents t k =
-          X.Contents.CA.copy contents t.X.Repo.contents "Contents" k
+        newies_commits >>= fun newies_commits ->
+        newies_nodes >>= fun newies_nodes ->
+        newies_contents >>= fun newies_contents ->
+        let newies_commits = List.rev_map (fun x -> `Commit x) newies_commits in
```

Comment:
`rev_append` reverses the `newies_commits` so `rev_map` is not necessary here

Reply:
unfortunatly `List.map` is not tail-recursive ; But I can add a `rev_append_map` function if you think that's useful.

Reply:
sorry, I missed that. I think we need to ensure that the order in which new commits are added is preserved. The `consume_newies` returns them in the reverse order, and `rev_append` reverse them too. So we should either do `List.rev_map |> List.rev` or the `rev_append_map` as you suggested.

Reply:
But now the `newies` are added to the next layer using `Repo.iter` (in https://github.com/mirage/irmin/pull/1124/files#diff-0656cbce221ba6715aa8e99bc19aea4bd241093462b5c68f4671223326349928R780) which should ensure that the reverse transitive order is preserved.

Reply:
yes, indeed.